src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Dependency/DependencyDag.cs

```diff
@@ -0,0 +1,797 @@
+            // reference. Then, we simply walk each reference's dependency
+            // chain and link the nodes to eachother as appropriate.
+            //
+
+            var allNodes = catalog.PlugIns
+                .Select(x => new Node(x))
+                .Concat(repository.GetAllReferences().Select(x => new Node(x)))
+                .ToSet();
+
+            var depToNode = allNodes.ToDictionary(
+                key => key.Target,
+                value => value);
+
+            foreach (var node in allNodes)
+            {
+                node.Target.Match(
+                    (CustomDataSourceReference x) =>
+                    {
+                        foreach (var p in x.TrackedProcessors.OfType<ICustomDataProcessorWithSourceParser>())
+                        {
+                            var deps = allNodes.Where(
+                                n => n.Target.Match(
+                                    r => false,
+                                    r => r is ISourceDataCookerReference sdcr &&
+                                            sdcr.Path.SourceParserId == p.SourceParserId))
+                                .Select(n => n.Target);
+                            foreach (var dep in deps)
+                            {
+                                //
+                                // The source cooker depends on the custom data processor,
+                                // as the custom data processor is providing the source
+                                // parser.
+                                //
+
+                                var depNode = depToNode[dep];
+                                depNode.AddDependency(node);
+                            }
+                        }
+                    },
+                    (IDataExtensionReference x) =>
+                    {
+                        foreach (var dep in GetAllDependencyReferences(repository, x))
+                        {
+                            Debug.Assert(depToNode.ContainsKey(dep));
+                            var depNode = depToNode[dep];
+                            node.AddDependency(depNode);
+                        }
+                    });
+            }
+
+            //
+            // Walk the entire DAG, and remove any dependencies that are also
+            // transitive dependencies. The walk starts from the roots.
+            // For example:
+            //
+            //     n                              n
+            //    / \                             |
+            //   d   |  will be simplified to     d
+            //    \  |                            |
+            //      d2                            d2
+            //
+            // This is because n needs d and d2, but d also needs d2, so
+            // fundamentally n only needs d.
+            //
+
+            var roots = allNodes.Where(x => x.Dependents.Count == 0).ToSet();
```

Comment:
>roots [](start = 16, length = 5)

not a big deal, but since we already have all the roots, could we just pass this into the constructor rather than recalculating?

Reply:
Yes, that's a good point. Thanks!